lib/rucio/client/s3client.py

```diff
@@ -0,0 +1,138 @@
+            else:
+                cred = get_s3_credentials()
+            REGION.set("s3client-%s" % account, cred)
+        try:
+            self.s3 = boto3.client("s3", **cred, verify=False)
+        except Exception as error:
+            self.logger(logging.ERROR, error)
+            raise error
+
+    def bucket_create(self, path):
+        """Create an S3 bucket.
+
+        param bucket_name: Bucket path, e.g. user.dquijote:/mybucket/
+        :return: True if bucket created, else False
+        """
+        # TODO: IAM policy structure
+        bucket_name, bucket_path = path.split(":")
+
+        try:
+            self.s3.head_bucket(Bucket=bucket_name)
+        except ClientError as error:
+            # TODO: catch 409 BucketAlreadyOwnedByYou and BucketAlreadyExists (bucket can be owned by another user)
+            if error.response["Error"]["Code"] == "404":
+                self.logger(logging.DEBUG, "Creating bucket %s" % bucket_name)
+                self.s3.create_bucket(Bucket=bucket_name)
+            elif error.response:
+                self.logger(logging.ERROR, error.response["Error"]["Message"])
+                return FAILURE
+
+        try:
+            self.s3.put_object(Bucket=bucket_name, Body="", Key=bucket_path)
+        except ClientError as error:
+            self.logger(logging.ERROR, error.response["Error"]["Message"])
+            return FAILURE
+        return SUCCESS
+
+    def bucket_upload(self, from_path, to_path):
+        """Upload a file/folder to an S3 bucket.
+
+        :param from_path: Path to the file/folder to upload
+        :param to_path: Bucket path, e.g. user.dquijote:/mybucket/file.ext or user.dquijote:/mybucket/folder/
+        :return: True if file/folder uploaded, else False
+        """
+        bucket_name, bucket_path = to_path.split(":")
+
+        if os.path.isdir(from_path):
+            for root, dirs, files in os.walk(from_path):
+                for file in files:
+                    with open(os.path.join(root, file), "rb") as f:
+                        # TODO: keep folder structure when uploading
+                        # TODO: add exceptions
+                        destination = (
+                            bucket_path + file if to_path.endswith("/") else bucket_path
+                        )
+                        try:
+                            self.s3.upload_fileobj(
+                                Fileobj=f, Bucket=bucket_name, Key=str(destination)
+                            )
+                        except ClientError as error:
+                            self.logger(logging.ERROR, error)
+                            return FAILURE
+                        return SUCCESS
+
+        else:
+            f = open(from_path, "rb")
+            # write file to folder_name/filename or file to filename if bucket_path is a file
+            destination = (
+                bucket_path + os.path.basename(from_path)
+                if bucket_path.endswith("/")
```

Comment:
is this proper python syntax?

Reply:
interesting, seems like that's what my formatter did, I tried using it to not manually fix all the flake8 errors (btw is this even allowed here, or should they be fixed manually?)